apps/src/blockly/addons/cdoFieldAngleDropdown.ts

```diff
@@ -1,118 +1,177 @@
-   * @returns A valid angle, or null if invalid.
+   * Get the direction from either the hardcoded setting or the direction field.
+   * @returns {string} The direction value.
+   * @private
    */
-  protected doClassValidation_(newValue: number = 0): number | null {
-    // The plugin can handle validation automatically, unless our block
-    // or XML serialization includes a set of valid options.
-    if (!this.configValues && !this.menuGeneratorValues) {
-      return super.doClassValidation_(newValue);
+  private getDirection(): string {
+    let direction = this.direction;
+    if (!direction && this.directionFieldName) {
+      const sourceBlock = this.getSourceBlock() as ExtendedBlockSvg;
+      direction = sourceBlock.getFieldValue(this.directionFieldName);
     }
+    return direction || 'turnRight';
+  }
 
-    // Get a valid angle value, then find the nearest custom option.
-    const validAngleValue = super.doClassValidation_(newValue);
-
-    // A field wouldn't normally have both a menu generator and xml config values.
-    // If both are present, we favor the config values as they come directly
-    // from the serialization.
-    const customValues = this.configValues || this.menuGeneratorValues;
-    const closestCustomValue = customValues.reduce(
-      (prev: number, curr: number) => {
-        return Math.abs(curr - validAngleValue) <
-          Math.abs(prev - validAngleValue)
-          ? curr
-          : prev;
-      }
+  /**
+   * Create the container element for the angle helper.
+   * @returns {HTMLDivElement} The angle helper container element.
+   * @private
+   */
+  private createAngleHelperContainer(): HTMLDivElement {
+    const angleHelperContainer = document.createElement('div');
+    Blockly.utils.dom.addClass(
+      angleHelperContainer,
+      'blocklyAngleHelperContainer'
     );
-    return closestCustomValue;
+    return angleHelperContainer;
   }
 
   /**
-   * Converts xml element into dropdown field
-   * @param {Element} element xml field
-   * @override
+   * Initialize the angle helper.
+   * @private
    */
-  fromXml(element: Element) {
-    // If the field xml contains a `config`, then we can use it to limit the valid selectable options.
-    this.config = element.getAttribute('config');
-    if (this.config) {
-      this.configValues = this.config.split(',').map(Number);
-      this.min_ = Math.min(...this.configValues);
-      this.max_ = Math.max(...this.configValues);
-    }
-    super.fromXml(element);
+  private initializeAngleHelper(): void {
+    const container = this.createAngleHelperContainer();
+    const sourceBlock = this.getSourceBlock() as Block;
+    this.angleHelper = new Blockly.AngleHelper(this.getDirection(), {
+      onUpdate: this.updateDropdownMenuOptions.bind(this),
+      snapPoints: this.getOptions().map(option => parseInt(option[1])),
+      arcColour: (sourceBlock as ExtendedBlockSvg)?.style.colourPrimary,
+      angle: parseInt(`${this.getValue()}`),
```

Comment:
why do we parseInt a stringified number? (Does getValue() return a number?)

Reply:
This was to satisfy TypeScript. In practice, `this.getValue()` actually returns a string, but Blockly says it could also be null. `parseInt` expect a string. In theory, `parseInt('null')` is okay as it returns `NaN`, but we can't really hit that case.